src/pytorch_lightning/utilities/migration/utils.py

```diff
@@ -0,0 +1,133 @@
+    """Applies Lightning version migrations to a checkpoint dictionary.
+
+    Note:
+        The migration happens in-place. We specifically avoid copying the dict to avoid memory spikes for large
+        checkpoints and objects that do not support being deep-copied.
+    """
+    ckpt_version = _get_version(checkpoint)
+    if LooseVersion(ckpt_version) > LooseVersion(pl.__version__):
+        rank_zero_warn(
+            f"The loaded checkpoint was produced with Lightning v{ckpt_version}, which is newer than your current"
+            f" Lightning version: v{pl.__version__}",
+            category=PossibleUserWarning,
+        )
+        return checkpoint, {}
+
+    index = migration_index()
+    applied_migrations = {}
+    for migration_version, migration_functions in index.items():
+        if not _should_upgrade(checkpoint, migration_version):
+            continue
+        for migration_function in migration_functions:
+            checkpoint = migration_function(checkpoint)
+
+        applied_migrations[migration_version] = [fn.__name__ for fn in migration_functions]
+
+    _set_version(checkpoint, pl.__version__)
+    return checkpoint, applied_migrations
+
+
+class pl_legacy_patch:
+    """Registers legacy artifacts (classes, methods, etc.) that were removed but still need to be included for
+    unpickling old checkpoints. The following patches apply.
+
+        1. ``pytorch_lightning.utilities.argparse._gpus_arg_default``: Applies to all checkpoints saved prior to
+           version 1.2.8. See: https://github.com/PyTorchLightning/pytorch-lightning/pull/6898
+        2. ``pytorch_lightning.utilities.argparse_utils``: A module that was deprecated in 1.2 and removed in 1.4,
+           but still needs to be available for import for legacy checkpoints.
+
+    Example:
+
+        with pl_legacy_patch():
+            torch.load("path/to/legacy/checkpoint.ckpt")
+    """
+
+    def __enter__(self) -> "pl_legacy_patch":
+        # `pl.utilities.argparse_utils` was renamed to `pl.utilities.argparse`
+        legacy_argparse_module = ModuleType("pytorch_lightning.utilities.argparse_utils")
+        sys.modules["pytorch_lightning.utilities.argparse_utils"] = legacy_argparse_module
+
+        # `_gpus_arg_default` used to be imported from these locations
+        legacy_argparse_module._gpus_arg_default = lambda x: x
+        pl.utilities.argparse._gpus_arg_default = lambda x: x
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        exc_traceback: Optional[TracebackType],
+    ) -> None:
+        if hasattr(pl.utilities.argparse, "_gpus_arg_default"):
+            delattr(pl.utilities.argparse, "_gpus_arg_default")
+        del sys.modules["pytorch_lightning.utilities.argparse_utils"]
+
+
+def _pl_migrate_checkpoint(checkpoint: _CHECKPOINT, checkpoint_path: _PATH) -> _CHECKPOINT:
```

Comment:
I'm a bit confused about the difference between this function and `migrate_checkpoint`. Can you clarify when which would be used? Could they be refactored so one calls the other?

Reply:
`upgrade_checkpoint.py` (the cli) uses the main `migrate_checkpoint` function. 
PL Trainer uses the `_pl_migrate_checkpoint` function, because it logs a message that is specific to the automatic ugprade. We don't want this message to appear for the upgrade script.

Reply:
> Could they be refactored so one calls the other?

It already is like this. The `_pl_migrate_checkpoint` calls `migrate_checkpoint`